src/System.Private.Interpreter/src/Internal/Runtime/Interpreter/ILImporter.Interpreter.cs

```diff
@@ -239,22 +265,963 @@ private void ImportLoadFloat(double value)
+                        switch (opcode)
+                        {
+                            case ILOpcode.shl:
+                                value = value << shiftBy;
+                                break;
+                            case ILOpcode.shr:
+                                value = value >> shiftBy;
+                                break;
+                            case ILOpcode.shr_un:
+                                value = (long)((ulong)value >> shiftBy);
+                                break;
+                            default:
+                                Debug.Assert(false);
+                                break;
+                        }
+
+                        _interpreter.EvaluationStack.Push(StackItem.FromInt64(value));
+                    }
+                    break;
+                case StackValueKind.NativeInt:
+                    {
+                        IntPtr value = op2.AsNativeInt();
+                        switch (opcode)
+                        {
+                            case ILOpcode.shl:
+                                value = new IntPtr(value.ToInt64() << shiftBy);
+                                break;
+                            case ILOpcode.shr:
+                                value = new IntPtr(value.ToInt64() >> shiftBy);
+                                break;
+                            case ILOpcode.shr_un:
+                                UIntPtr uintPtr = (UIntPtr)value.ToPointer();
+                                value = new IntPtr((long)(uintPtr.ToUInt64() >> shiftBy));
+                                break;
+                            default:
+                                Debug.Assert(false);
+                                break;
+                        }
+
+                        _interpreter.EvaluationStack.Push(StackItem.FromNativeInt(value));
+                    }
+                    break;
+                case StackValueKind.Unknown:
+                case StackValueKind.Float:
+                case StackValueKind.ByRef:
+                case StackValueKind.ObjRef:
+                case StackValueKind.ValueType:
+                default:
+                    ThrowHelper.ThrowInvalidProgramException();
+                    break;
+            }
         }
 
         private void ImportCompareOperation(ILOpcode opcode)
         {
-            throw new NotImplementedException();
+            bool result = default(bool);
+            StackItem op1 = PopWithValidation();
+            StackItem op2 = PopWithValidation();
+
+            switch (op1.Kind)
+            {
+                case StackValueKind.Int32:
+                    {
+                        int val1 = op1.AsInt32();
+
+                        if (op2.Kind == StackValueKind.Int32 || op2.Kind == StackValueKind.NativeInt)
+                        {
+                            long val2 = op2.Kind == StackValueKind.Int32 ? op2.AsInt32() : op2.AsNativeInt().ToInt64();
```

Comment:
Is the zero / sign extension always going to be right here for native int?

Consider: 32-bit platform, the stack has NativeInt(0xFFFFFFFF) and int32(1), and we are doing unsigned comparison.

The correct answer should be  `NativeInt(0xFFFFFFFF)` > `int32(1)`.

But what is going to happen here is: `op2.AsNativeInt().ToInt64()` will end up to be `-1` and we will get wrong answer.

Reply:
I think this can be fixed nicely using the trick I have mentioned in the other comment:
```
// StackValueKind is carefully ordered to make this work (assuming the IL is valid)
StackValueKind kind = (op1.Kind > op2.Kind) ? op1.Kind : op2.Kind;
```

Reply:
It may be a new set of methods in addition to the existing `As*` methods.

Reply:
Or perhaps an optional boolean argument that specifies whether to assert and is set to true by default?

Reply:
> Or perhaps an optional boolean argument that specifies whether to assert and is set to true by default?

Something like `op2.AsNativeIntUnchecked()` might look better than `op2.AsNativeInt(assert: false)`.